File: api/src/main/java/com/medrec/services/AppointmentsService.java
package com.medrec.services;

import com.medrec.dtos.appointments.appointment.*;
import com.medrec.dtos.appointments.diagnosis.CreateDiagnosisDTO;
import com.medrec.dtos.appointments.diagnosis.DiagnosisDTO;
import com.medrec.dtos.appointments.diagnosis.UpdateDiagnosisDTO;
import com.medrec.dtos.appointments.icd.CreateIcdDTO;
import com.medrec.dtos.appointments.icd.IcdDTO;
import com.medrec.dtos.appointments.icd.UpdateIcdDTO;
import com.medrec.dtos.appointments.sick_leave.CreateSickLeaveDTO;
import com.medrec.dtos.appointments.sick_leave.SickLeaveDTO;
import com.medrec.dtos.appointments.sick_leave.UpdateSickLeaveDTO;
import com.medrec.gateways.AppointmentsGateway;
import com.medrec.grpc.appointments.Appointments;
import com.medrec.utils.Utils;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Logger;

@Service
public class AppointmentsService {
    private final Logger logger = Logger.getLogger(AppointmentsService.class.getName());

    private final AppointmentsGateway appointmentsGateway;

    public AppointmentsService(AppointmentsGateway appointmentsGateway) {
        this.appointmentsGateway = appointmentsGateway;
    }

    public AppointmentDTO createAppointment(CreateAppointmentDTO dto) throws RuntimeException {
        this.logger.info("Creating new appointment");

        Appointments.CreateAppointmentRequest request = Appointments.CreateAppointmentRequest.newBuilder()
            .setDate(dto.getDate())
            .setTime(dto.getTime())
            .setDoctorId(dto.getDoctorId())
            .setPatientId(dto.getPatientId())
            .build();

        try {
            Appointments.Appointment response = this.appointmentsGateway.createAppointment(request);
            AppointmentDTO appointmentDTO = Utils.getDTOFromAppointmentsGrpc(response);

            this.logger.info("New appointment created with id: " + appointmentDTO.getId());
            return appointmentDTO;
        } catch (RuntimeException e) {
            this.logger.warning("Could not create new appointment");
            throw e;
        }
    }

    public AppointmentDTO getAppointmentById(int id) {
        this.logger.info("Retrieving appointment with id: " + id);

        try {
            Appointments.Appointment appointment = this.appointmentsGateway.getAppointmentById(id);
            AppointmentDTO appointmentDTO = Utils.getDTOFromAppointmentsGrpc(appointment);

            this.logger.info("Appointment retrieved with id: " + appointmentDTO.getId());
            return appointmentDTO;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve appointment with id: " + id);
            throw e;
        }
    }

    public List<AppointmentDTO> getAllAppointments() {
        this.logger.info("Retrieving all appointments");

        try {
            Appointments.AppointmentsList appointments = this.appointmentsGateway.getAllAppointments();
            List<AppointmentDTO> appointmentDTOs = appointments.getAppointmentsList().stream()
                .map(Utils::getDTOFromAppointmentsGrpc)
                .toList();

            this.logger.info(appointmentDTOs.size() + "appointments retrieved");
            return appointmentDTOs;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all appointments");
            throw e;
        }
    }

    public List<AppointmentDTO> getAllByPatientEmail(String email) {
        this.logger.info("Retrieving all appointments for patient with email: " + email);

        try {
            Appointments.AppointmentsList appointments = this.appointmentsGateway.getAllByPatientEmail(email);
            List<AppointmentDTO> appointmentDTOs = appointments.getAppointmentsList().stream()
                .map(Utils::getDTOFromAppointmentsGrpc)
                .toList();

            this.logger.info(appointmentDTOs.size() + "appointments retrieved for patient with email: " + email);
            return appointmentDTOs;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all appointments for patient with email: " + email);
            throw e;
        }
    }

    public List<AppointmentDTO> getAllByPatientId(int id) {
        this.logger.info("Retrieving all appointments for patient with id: " + id);

        try {
            Appointments.AppointmentsList appointments = this.appointmentsGateway.getAllByPatientId(id);
            List<AppointmentDTO> appointmentDTOs = appointments.getAppointmentsList().stream()
                .map(Utils::getDTOFromAppointmentsGrpc)
                .toList();

            this.logger.info(appointmentDTOs.size() + "appointments retrieved for patient with id: " + id);
            return appointmentDTOs;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all appointments for patient with id: " + id);
            throw e;
        }
    }

    public List<AppointmentDTO> getAllByDoctorEmail(String email) {
        this.logger.info("Retrieving all appointments for doctor with email: " + email);

        try {
            Appointments.AppointmentsList appointments = this.appointmentsGateway.getAllByDoctorEmail(email);
            List<AppointmentDTO> appointmentDTOs = appointments.getAppointmentsList().stream()
                .map(Utils::getDTOFromAppointmentsGrpc)
                .toList();

            this.logger.info(appointmentDTOs.size() + " appointments retrieved for doctor with email: " + email);
            return appointmentDTOs;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all appointments for doctor with email: " + email);
            throw e;
        }
    }

    public List<AppointmentDTO> getAllByDoctorId(int id) {
        this.logger.info("Retrieving all appointments for doctor with id: " + id);

        try {
            Appointments.AppointmentsList appointments = this.appointmentsGateway.getAllByDoctorId(id);
            List<AppointmentDTO> appointmentDTOs = appointments.getAppointmentsList().stream()
                .map(Utils::getDTOFromAppointmentsGrpc)
                .toList();

            this.logger.info(appointmentDTOs.size() + "appointments retrieved for doctor with id: " + id);
            return appointmentDTOs;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all appointments for doctor with id: " + id);
            throw e;
        }
    }

    public AppointmentDTO updateAppointment(UpdateAppointmentDTO dto) throws RuntimeException {
        this.logger.info("Updating appointment with id " + dto.getId());

        try {
            Appointments.Appointment updatedAppointment = this.appointmentsGateway.updateAppointment(dto.safeConvertToGrpcRequest());
            AppointmentDTO appointmentDTO = Utils.getDTOFromAppointmentsGrpc(updatedAppointment);

            this.logger.info("Appointment updated with id: " + appointmentDTO.getId());
            return appointmentDTO;
        } catch (RuntimeException e) {
            this.logger.warning("Could not update appointment with id: " + dto.getId());
            throw e;
        }
    }

    public void deleteAppointmentById(int id) throws RuntimeException {
        this.logger.info("Deleting appointment with id: " + id);

        try {
            this.appointmentsGateway.deleteAppointment(id);
            this.logger.info("Appointment deleted with id: " + id);
        } catch (RuntimeException e) {
            this.logger.warning("Could not delete appointment with id: " + id);
            throw e;
        }
    }

    public void cascadeDeletePatientAppointments(int patientId) throws RuntimeException {
        this.logger.info("Deleting all appointments for patient with id: " + patientId);

        try {
            this.appointmentsGateway.cascadeDeletePatientAppointments(patientId);
            this.logger.info("All appointments for patient with id: " + patientId + " deleted");
        } catch (RuntimeException e) {
            this.logger.warning("Could not delete all appointments for patient with id: " + patientId);
            throw e;
        }
    }

    public void cascadeDeleteDoctorAppointments(int doctorId) throws RuntimeException {
        this.logger.info("Deleting all appointments for doctor with id: " + doctorId);

        try {
            this.appointmentsGateway.cascadeDeleteDoctorAppointments(doctorId);
            this.logger.info("All appointments for doctor with id: " + doctorId + " deleted");
        } catch (RuntimeException e) {
            this.logger.warning("Could not delete all appointments for doctor with id: " + doctorId);
            throw e;
        }
    }
    public SickLeaveDTO createSickLeave(CreateSickLeaveDTO dto) throws RuntimeException {
        this.logger.info("Creating new Sick Leave entity");

        Appointments.CreateSickLeaveRequest request = Appointments.CreateSickLeaveRequest.newBuilder()
            .setDate(dto.getDate())
            .setDaysOfLeave(dto.getNumberOfDays())
            .build();

        try {
            Appointments.SickLeave sickLeave = this.appointmentsGateway.createSickLeave(request);
            SickLeaveDTO sickLeaveDto = Utils.getDtoFromSickLeaveGrpc(sickLeave);

            this.logger.info("Created Sick Leave entity with id: " + sickLeaveDto.getId());
            return sickLeaveDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not create new Sick Leave entity");
            throw e;
        }
    }

    public SickLeaveDTO getSickLeaveById(int id) throws RuntimeException {
        this.logger.info("Retrieving Sick Leave entity with id: " + id);

        try {
            Appointments.SickLeave sickLeave = this.appointmentsGateway.getSickLeaveById(id);
            SickLeaveDTO sickLeaveDTO = Utils.getDtoFromSickLeaveGrpc(sickLeave);

            this.logger.info("Retrieved Sick Leave entity with id: " + id);
            return sickLeaveDTO;
        } catch (RuntimeException e) {
            this.logger.info("Error retrieving Sick Leave entity: " + e.getMessage());
            throw e;
        }
    }

    public List<SickLeaveDTO> getAllSickLeaveEntities() throws RuntimeException {
        this.logger.info("Retrieving all doctors");

        try {
            Appointments.SickLeaveEntitiesList entities = this.appointmentsGateway.getAllSickLeaveEntities();
            List<SickLeaveDTO> sickLeaveDtos = new ArrayList<>();

            entities.getSickLeaveEntitiesList().forEach(entity -> {
                sickLeaveDtos.add(Utils.getDtoFromSickLeaveGrpc(entity));
            });

            this.logger.info("Retrieved " + sickLeaveDtos.size() + " Sick Leave entities");
            return sickLeaveDtos;
        } catch (RuntimeException e) {
            this.logger.info("Error retrieving Sick Leave entities: " + e.getMessage());
            throw e;
        }
    }

    public SickLeaveDTO updateSickLeave(UpdateSickLeaveDTO dto) throws RuntimeException {
        this.logger.info("Updating sick leave entity with id " + dto.getId());

        try {
            Appointments.SickLeave sickLeave = this.appointmentsGateway.updateSickLeave(dto.safeConvertToGrpcRequest());
            SickLeaveDTO sickLeaveDto = Utils.getDtoFromSickLeaveGrpc(sickLeave);

            this.logger.info("Successfully update sick leave with id: " + sickLeaveDto.getId());
            return sickLeaveDto;
        } catch (RuntimeException e) {
            this.logger.info("Error updating Sick Leave: " + e.getMessage());
            throw e;
        }
    }

    public void deleteSickLeaveById(int id) throws RuntimeException {
        this.logger.info("Deleting Sick Leave entity with id: " + id);

        try {
            this.appointmentsGateway.deleteSickLeave(id);
            this.logger.info("Deleted Sick Leave entity with id: " + id);
        } catch (RuntimeException e) {
            this.logger.info("Error deleting sick leave entity: " + e.getMessage());
            throw e;
        }
    }

    public IcdDTO createIcd(CreateIcdDTO dto) throws RuntimeException {
        this.logger.info("Creating new ICD");

        Appointments.CreateIcdRequest request = Appointments.CreateIcdRequest.newBuilder()
            .setCode(dto.getCode())
            .setDescription(dto.getDescription())
            .build();

        try {
            Appointments.Icd icd = this.appointmentsGateway.createIcd(request);
            IcdDTO icdDto = Utils.getDtoFromIcdGrpc(icd);

            this.logger.info("Created ICD with id: " + icdDto.getId());
            return icdDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not create new ICD");
            throw e;
        }
    }

    public IcdDTO getIcdById(int id) throws RuntimeException {
        this.logger.info("Retrieving ICD with id: " + id);

        try {
            Appointments.Icd icd = this.appointmentsGateway.getIcdById(id);
            IcdDTO icdDto = Utils.getDtoFromIcdGrpc(icd);

            this.logger.info("Retrieved ICD with id: " + id);
            return icdDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve ICD with id: " + id);
            throw e;
        }
    }

    public List<IcdDTO> getAllIcds() throws RuntimeException {
        this.logger.info("Retrieving all ICDs");

        try {
            Appointments.IcdEntitiesList icds = this.appointmentsGateway.getAllIcdEntities();
            List<IcdDTO> icdDtos = new ArrayList<>();

            icds.getIcdEntitiesList().forEach(icd -> {
                icdDtos.add(Utils.getDtoFromIcdGrpc(icd));
            });

            this.logger.info("Retrieved " + icdDtos.size() + " ICDs");
            return icdDtos;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all ICDs");
            throw e;
        }
    }

    public IcdDTO updateIcd(UpdateIcdDTO dto) throws RuntimeException {
        this.logger.info("Updating ICD with id " + dto.getId());

        try {
            Appointments.Icd icd = this.appointmentsGateway.updateIcd(dto.safeConvertToGrpcRequest());
            IcdDTO icdDto = Utils.getDtoFromIcdGrpc(icd);

            this.logger.info("Successfully updated ICD with id: " + icdDto.getId());
            return icdDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not update ICD with id: " + dto.getId());
            throw e;
        }
    }

    public void deleteIcdById(int id) throws RuntimeException {
        this.logger.info("Deleting ICD with id: " + id);

        try {
            this.appointmentsGateway.deleteIcd(id);
            this.logger.info("Deleted ICD with id: " + id);
        } catch (RuntimeException e) {
            this.logger.warning("Could not delete ICD with id: " + id);
            throw e;
        }
    }

    public DiagnosisDTO createDiagnosis(CreateDiagnosisDTO dto) throws RuntimeException {
        this.logger.info("Creating new diagnosis");

        try {
            Appointments.Diagnosis diagnosis = this.appointmentsGateway.createDiagnosis(dto.safeConvertToGrpcRequest());
            DiagnosisDTO diagnosisDto = Utils.getDtoFromDiagnosisGrpc(diagnosis);

            this.logger.info("Created diagnosis with id: " + diagnosisDto.getId());
            return diagnosisDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not create new diagnosis");
            throw e;
        }
    }

    public DiagnosisDTO getDiagnosisById(int id) throws RuntimeException {
        this.logger.info("Retrieving diagnosis by id: " + id);

        try {
            Appointments.Diagnosis diagnosis = this.appointmentsGateway.getDiagnosisById(id);
            DiagnosisDTO diagnosisDto = Utils.getDtoFromDiagnosisGrpc(diagnosis);

            this.logger.info("Retrieved diagnosis with id: " + diagnosisDto.getId());
            return diagnosisDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve diagnosis with id: " + id);
            throw e;
        }
    }

    public List<DiagnosisDTO> getAllDiagnoses() throws RuntimeException {
        this.logger.info("Retrieving all diagnoses");

        try {
            Appointments.DiagnosesList diagnoses = this.appointmentsGateway.getAllDiagnoses();
            List<DiagnosisDTO> diagnosesDtos = new ArrayList<>();

            diagnoses.getDiagnosesList().forEach(diagnosis -> {
                diagnosesDtos.add(Utils.getDtoFromDiagnosisGrpc(diagnosis));
            });

            this.logger.info("Retrieved " + diagnosesDtos.size() + " diagnoses");
            return diagnosesDtos;
        } catch (RuntimeException e) {
            this.logger.warning("Could not retrieve all diagnoses: " + e.getMessage());
            throw e;
        }
    }

    public DiagnosisDTO updateDiagnosis(UpdateDiagnosisDTO dto) throws RuntimeException {
        this.logger.info("Updating diagnosis with id: " + dto.getId());

        try {
            Appointments.Diagnosis diagnosis = this.appointmentsGateway.updateDiagnosis(dto.safeConvertToGrpcRequest());
            DiagnosisDTO diagnosisDto = Utils.getDtoFromDiagnosisGrpc(diagnosis);

            this.logger.info("Successfully updated diagnosis with id: " + diagnosisDto.getId());
            return diagnosisDto;
        } catch (RuntimeException e) {
            this.logger.warning("Could not update diagnosis: " + e.getMessage());
            throw e;
        }
    }

    public void deleteDiagnosisById(int id) throws RuntimeException {
        this.logger.info("Deleting diagnosis with id: " + id);

        try {
            this.appointmentsGateway.deleteDiagnosis(id);
            this.logger.info("Deleted diagnosis with id: " + id);
        } catch (RuntimeException e) {
            this.logger.warning("Could not delete diagnosis with id: " + id);
            throw e;
        }
    }

    public List<IcdDTO> startAppointmentFetchIcds(StartAppointmentDTO dto) throws RuntimeException {
        int appointmentId = dto.getAppointmentId();
        int doctorId = dto.getDoctorId();

        this.logger.info("Starting appointment with id: " + appointmentId);

        try {
            Appointments.StartAppointmentFetchIcdsRequest request = Appointments.StartAppointmentFetchIcdsRequest.newBuilder()
                .setAppointmentId(appointmentId)
                .setDoctorId(doctorId)
                .build();

            Appointments.IcdEntitiesList icdEntitiesList = this.appointmentsGateway.startAppointmentFetchIcds(request);

            List<IcdDTO> icdDtos = new ArrayList<>();
            icdEntitiesList.getIcdEntitiesList().forEach(icd -> {
                icdDtos.add(Utils.getDtoFromIcdGrpc(icd));
            });

            this.logger.info("Retrieved " + icdDtos.size() + " icds");
            return icdDtos;
        } catch (RuntimeException e) {
            this.logger.warning("Could not start appointment with id: " + appointmentId);
            throw e;
        }
    }

    public void finishAppointmentAddDiagnosis(FinishAppointmentDTO dto) throws RuntimeException {
        int appointmentId = dto.getAppointmentId();
        this.logger.info("Finishing appointment with id: " + appointmentId);

        try {
            CreateDiagnosisDTO createDiagnosisDto = new CreateDiagnosisDTO(
                dto.getTreatmentDescription(),
                dto.getIcdId(),
                dto.getSickLeaveDate(),
                dto.getSickLeaveDays()
            );
            Appointments.FinishAppointmentAddDiagnosisRequest request = Appointments.FinishAppointmentAddDiagnosisRequest.newBuilder()
                .setAppointmentId(appointmentId)
                .setDiagnosis(createDiagnosisDto.safeConvertToGrpcRequest())
                .build();

            this.appointmentsGateway.finishAppointmentAddDiagnosis(request);
        } catch (RuntimeException e) {
            this.logger.warning("Could not finish appointment with id: " + appointmentId);
            throw e;
        }
    }
}

File: frontend/src/app/services/appointments.service.ts
import { Injectable } from '@angular/core';
import {environment} from '../../environments/environment';
import {HttpClient, HttpResponse} from '@angular/common/http';
import {AppointmentDto} from '../common/dtos/appointment.dto';
import {firstValueFrom, Observable} from 'rxjs';
import {ApiResponse} from '../common/interfaces/api.response';
import {DiagnosisDto} from '../common/dtos/diagnosis.dto';
import {IcdDto} from '../common/dtos/icd.dto';
import {SickLeaveDto} from '../common/dtos/sick-leave.dto';

@Injectable({
  providedIn: 'root'
})
export class AppointmentsService {
  private readonly apiUrl: string = environment.apiGateway

  constructor(
    private httpClient: HttpClient,
  ) { }

  public async getAllAppointments(): Promise<AppointmentDto[]> {
    return firstValueFrom(this.httpClient.get<AppointmentDto[]>(`${this.apiUrl}/appointments/list/all`));
  }

  public createAppointment(date: string, time: string, patientId: number, doctorId: number): Observable<HttpResponse<ApiResponse>> {
    let body: any = {
      date: date,
      time: time,
      patientId: patientId,
      doctorId: doctorId
    };

    return this.httpClient.post<ApiResponse>(`${this.apiUrl}/appointments/create`, body, {observe: 'response'});
  }

  public updateAppointment(id: number, status: string | null, diagnosisId: number | null): Observable<HttpResponse<ApiResponse>> {
    let body: any = {};
    body.id = id;

    if (status) {
      body.status = status;
    }

    if (diagnosisId) {
      body.diagnosisId = diagnosisId;
    }

    return this.httpClient.put<ApiResponse>(`${this.apiUrl}/appointments/update`, body, {observe: 'response'});
  }

  public deleteAppointment(id: number): Observable<HttpResponse<ApiResponse>> {
    return this.httpClient.delete<ApiResponse>(`${this.apiUrl}/appointments/delete?id=${id}`, {observe: 'response'});
  }

  public async getAllDiagnoses(): Promise<DiagnosisDto[]> {
    return firstValueFrom(this.httpClient.get<DiagnosisDto[]>(`${this.apiUrl}/diagnosis/list/all`));
  }

  public createDiagnosis(treatmentDescription: string, icdId: number, sickLeaveDate: number | null, sickLeaveDays: number | null): Observable<HttpResponse<ApiResponse>> {
    let body: any = {};
    body.treatmentDescription = treatmentDescription;
    body.icdId = icdId;

    if (sickLeaveDate && sickLeaveDays) {
      body.sickLeaveDate = sickLeaveDate;
      body.sickLeaveDays = sickLeaveDays;
    }

    return this.httpClient.post<ApiResponse>(`${this.apiUrl}/diagnosis/create`, body, {observe: 'response'});
  }

  public updateDiagnosis(id: number, treatmentDescription: string, icdId: number, sickLeaveDate: number | null, sickLeaveDays: number | null): Observable<HttpResponse<ApiResponse>> {
    let body: any = {}
    body.id = id;

    if (treatmentDescription) {
      body.treatmentDescription = treatmentDescription;
    }

    if (icdId) {
      body.icdId = icdId;
    }

    if (sickLeaveDate && sickLeaveDays) {
      body.sickLeaveDate = sickLeaveDate;
      body.sickLeaveDays = sickLeaveDays;
    }

    return this.httpClient.put<ApiResponse>(`${this.apiUrl}/diagnosis/update`, body, {observe: 'response'});
  }

  public deleteDiagnosis(id: number): Observable<HttpResponse<ApiResponse>> {
    return this.httpClient.delete<ApiResponse>(`${this.apiUrl}/diagnosis/delete?id=${id}`, {observe: 'response'});
  }

  public async getAllIcdEntries(): Promise<IcdDto[]> {
    return firstValueFrom(this.httpClient.get<IcdDto[]>(`${this.apiUrl}/icd/list/all`));
  }

  public createIcd(icdCode: string, icdDescription: string): Observable<HttpResponse<ApiResponse>> {
    const body: any = {
      code: icdCode,
      description: icdDescription
    }

    return this.httpClient.post<ApiResponse>(`${this.apiUrl}/icd/create`, body, {observe: 'response'});
  }

  public updateIcd(id: number, icdCode: string | null, icdDescription: string | null): Observable<HttpResponse<ApiResponse>> {
    let body: any = {}
    body.id = id;

    if (icdCode) {
      body.code = icdCode;
    }

    if (icdDescription) {
      body.description = icdDescription;
    }

    return this.httpClient.put<ApiResponse>(`${this.apiUrl}/icd/update`, body, {observe: 'response'});
  }

  public deleteIcd(id: number): Observable<HttpResponse<ApiResponse>> {
    return this.httpClient.delete<ApiResponse>(`${this.apiUrl}/icd/delete?id=${id}`, {observe: 'response'});
  }

  public async getAllSickLeaveEntities(): Promise<SickLeaveDto[]> {
    return firstValueFrom(this.httpClient.get<SickLeaveDto[]>(`${this.apiUrl}/sick-leave/list/all`));
  }

  public createSickLeave(date: string, days: number): Observable<HttpResponse<ApiResponse>> {
    const body = {
      date: date,
      numberOfDays: days
    }

    return this.httpClient.post<ApiResponse>(`${this.apiUrl}/sick-leave/create`, body, {observe: 'response'});
  }

  public updateSickLeave(id: number, date: string | null, days: number | null): Observable<HttpResponse<ApiResponse>> {
    let body: any = {}
    body.id = id;

    if (date) {
      body.date = date;
    }

    if (days) {
      body.numberOfDays = days;
    }

    return this.httpClient.put<ApiResponse>(`${this.apiUrl}/sick-leave/update`, body, {observe: 'response'});
  }

  public deleteSickLeave(id: number): Observable<HttpResponse<ApiResponse>> {
    return this.httpClient.delete<ApiResponse>(`${this.apiUrl}/sick-leave/delete?id=${id}`, {observe: 'response'});
  }

  public startAppointment(appointmentId: number, doctorId: number): Observable<HttpResponse<IcdDto[]>> {
    const body = {
      appointmentId: appointmentId,
      doctorId: doctorId
    }

    return this.httpClient.post<IcdDto[]>(`${this.apiUrl}/appointments/start`, body, {observe: 'response'});
  }

  public finishAppointment(
    appointmentId: number,
    treatmentDescription: string,
    icdId: number,
    sickLeaveDate: string | null,
    sickLeaveDays: number | null
  ): Observable<HttpResponse<ApiResponse>> {
    let body: any = {
      appointmentId: appointmentId,
      treatmentDescription: treatmentDescription,
      icdId: icdId
    }

    if (sickLeaveDate) {
      body.sickLeaveDate = sickLeaveDate;
    }

    if (sickLeaveDays) {
      body.sickLeaveDays = sickLeaveDays;
    }

    return this.httpClient.post<ApiResponse>(`${this.apiUrl}/appointments/finish`, body, {observe: "response"});
  }
}
